tests/Unit/Evolution/Systems/ScalarWave/Test_Constraints.cpp

```diff
@@ -0,0 +1,344 @@
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+#include "tests/Unit/Pypp/CheckWithRandomValues.hpp"
+#include "tests/Unit/Pypp/SetupLocalPythonEnvironment.hpp"
+
+// IWYU pragma: no_forward_declare Tensor
+// IWYU pragma: no_forward_declare Variables
+// IWYU pragma: no_forward_declare Tags::deriv
+
+namespace {
+// Test the return-by-value two-index constraint function using random values
+template <size_t Dim>
+void test_two_index_constraint_random(
+    const DataVector& used_for_size) noexcept {
+  pypp::check_with_random_values<1>(
+      static_cast<tnsr::ij<DataVector, Dim, Frame::Inertial> (*)(
+          const tnsr::ij<DataVector, Dim, Frame::Inertial>&)>(
+          &ScalarWave::two_index_constraint<Dim>),
+      "TestFunctions", "two_index_constraint", {{{-10.0, 10.0}}}, used_for_size,
+      1.0e-12);
+}
+
+// Test the return-by-reference two-index constraint
+// by comparing to a time-independent analytic solution (e.g. Kerr-Schild)
+template <typename Solution>
+void test_two_index_constraint_analytic(
+    const Solution& solution, const size_t grid_size_each_dimension,
+    const std::array<double, 3>& lower_bound,
+    const std::array<double, 3>& upper_bound,
+    const double error_tolerance) noexcept {
+  // Shorter names for tags.
+  constexpr size_t dim = 3;
+  using VariablesTags =
+      tmpl::list<ScalarWave::Pi, ScalarWave::Phi<dim>, ScalarWave::Psi>;
+
+  // Check vs. time-independent analytic solution
+  // Set up grid
+  const size_t data_size = pow<3>(grid_size_each_dimension);
+  Mesh<3> mesh{grid_size_each_dimension, Spectral::Basis::Legendre,
+               Spectral::Quadrature::GaussLobatto};
+
+  using Affine = domain::CoordinateMaps::Affine;
+  using Affine3D =
+      domain::CoordinateMaps::ProductOf3Maps<Affine, Affine, Affine>;
+  const auto coord_map =
+      domain::make_coordinate_map<Frame::Logical, Frame::Inertial>(Affine3D{
+          Affine{-1.0, 1.0, lower_bound[0], upper_bound[0]},
+          Affine{-1.0, 1.0, lower_bound[1], upper_bound[1]},
+          Affine{-1.0, 1.0, lower_bound[2], upper_bound[2]},
+      });
+
+  // Set up coordinates
+  const auto x_logical = logical_coordinates(mesh);
+  const auto x = coord_map(x_logical);
+  // Arbitrary time for time-independent solution.
+  const double t = 0.;
+
+  // Evaluate analytic solution
+  const auto vars = solution.variables(
+      x, t,
+      tmpl::list<ScalarWave::Pi, ScalarWave::Phi<dim>, ScalarWave::Psi>{});
+
+  // Compute derivatives d_psi, d_phi, d_pi, and d_gauge_function numerically
+  Variables<VariablesTags> sw_vars(data_size);
+  get<ScalarWave::Psi>(sw_vars) = get<ScalarWave::Psi>(vars);
+  get<ScalarWave::Pi>(sw_vars) = get<ScalarWave::Pi>(vars);
+  get<ScalarWave::Phi<dim>>(sw_vars) = get<ScalarWave::Phi<dim>>(vars);
+
+  // Compute numerical derivatives of psi,pi,phi,H
+  const auto sw_derivs =
+      partial_derivatives<VariablesTags, VariablesTags, dim, Frame::Inertial>(
```

Comment:
- You can omit at least some of these template parameters
- if I understand correctly you only need the derivs of phi, so no need to compute the other derivs

Reply:
changed